src/flags.rs

```diff
@@ -0,0 +1,192 @@
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::mem;
+
+// Creates vector of strings, Vec<String>
+#[cfg(test)]
+macro_rules! svec {
+    ($($x:expr),*) => (vec![$($x.to_string()),*]);
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DenoFlags {
+  pub help: bool,
+  pub log_debug: bool,
+  pub version: bool,
+  pub reload: bool,
+  pub allow_write: bool,
+  pub allow_net: bool,
+}
+
+pub fn print_usage() {
+  println!(
+    "Usage: deno script.js
+
+-help         Print this message.
+-version      Print the version.
+-reload       Reload cached remote resources.
+-allow-write  Allow file system write access.
+-allow-net    Allow network access.
+-DEBUG        Log debug output."
+  );
+}
+
+// Parses flags for deno. This does not do v8_set_flags() - call that separately.
+pub fn set_flags(args: Vec<String>) -> (DenoFlags, Vec<String>) {
+  let mut flags = DenoFlags {
+    help: false,
+    version: false,
+    reload: false,
+    log_debug: false,
+    allow_write: false,
+    allow_net: false,
+  };
+  let mut rest = Vec::new();
+  for a in &args {
+    match a.as_str() {
+      "-h" | "-help" | "--help" => flags.help = true,
+      "-D" | "-DEBUG" | "--DEBUG" => flags.log_debug = true,
+      "-v" | "-version" | "--version" => flags.version = true,
+      "-r" | "-reload" | "--reload" => flags.reload = true,
+      "-allow-write" | "--allow-write" => flags.allow_write = true,
+      "-allow-net" | "--allow-net" => flags.allow_net = true,
+      _ => rest.push(a.clone()),
+    }
+  }
+
+  return (flags, rest);
+}
+
+#[test]
+fn test_set_flags_1() {
+  let (flags, rest) = set_flags(svec!["deno", "-version"]);
+  assert!(rest == svec!["deno"]);
+  assert!(
+    flags == DenoFlags {
+      help: false,
+      log_debug: false,
+      version: true,
+      reload: false,
+      allow_write: false,
+      allow_net: false,
+    }
+  );
+}
+
+#[test]
+fn test_set_flags_2() {
+  let (flags, rest) = set_flags(svec!["deno", "-r", "-D", "script.ts"]);
+  assert!(rest == svec!["deno", "script.ts"]);
+  assert!(
+    flags == DenoFlags {
+      help: false,
+      log_debug: true,
+      version: false,
+      reload: true,
+      allow_write: false,
+      allow_net: false,
+    }
+  );
+}
+
+#[test]
+fn test_set_flags_3() {
+  let (flags, rest) =
+    set_flags(svec!["deno", "-r", "script.ts", "-allow-write"]);
+  assert!(rest == svec!["deno", "script.ts"]);
+  assert!(
+    flags == DenoFlags {
+      help: false,
+      log_debug: false,
+      version: false,
+      reload: true,
+      allow_write: true,
+      allow_net: false,
+    }
+  );
+}
+
+// Returns args passed to V8, followed by args passed to JS
+// TODO Rename to v8_set_flags_preprocess
+fn parse_core_args(args: Vec<String>) -> (Vec<String>, Vec<String>) {
+  let mut rest = vec![];
+
+  // Filter out args that shouldn't be passed to V8
+  let mut args: Vec<String> = args
+    .into_iter()
+    .filter(|arg| {
+      if arg.as_str() == "--help" {
+        rest.push(arg.clone());
```

Comment:
It's just a few bytes so it doesn't matter here - generally speaking though it's not ideal and zero-cost if you copy the item and then throw away the original.
The way to do this is to use the iterator .partition() method instead of .filter().
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition